Stripe/STPPaymentMethodParams.m

```diff
@@ -7,25 +7,29 @@
 //
 
 #import "STPPaymentMethodParams.h"
-#import "STPPaymentMethod.h"
+#import "STPPaymentMethod+Private.h"
 
 @implementation STPPaymentMethodParams
 
 @synthesize additionalAPIParameters = _additionalAPIParameters;
 
 + (STPPaymentMethodParams *)paramsWithCard:(STPPaymentMethodCardParams *)card billingDetails:(STPPaymentMethodBillingDetails *)billingDetails metadata:(NSDictionary<NSString *,NSString *> *)metadata {
-    STPPaymentMethodParams *params = [STPPaymentMethodParams new];
-    params.type = STPPaymentMethodParamsTypeCard;
+    STPPaymentMethodParams *params = [self new];
+    params.type = STPPaymentMethodTypeCard;
     params.card = card;
     params.billingDetails = billingDetails;
     params.metadata = metadata;
     return params;
 }
 
-- (NSString *)apiStringType {
-    switch (self.type) {
-        case STPPaymentMethodParamsTypeCard:
-            return @"card";
+- (STPPaymentMethodType)type {
+    return [STPPaymentMethod typeFromString:self.rawTypeString];
+}
+
+- (void)setType:(STPPaymentMethodType)type {
+    // If setting unknown and we're already unknown, don't want to override raw value
+    if (type != self.type) {
+        self.rawTypeString = [STPPaymentMethod stringFromType:type];
```

Comment:
It looks like setting `type = .unknown` will set `rawTypeString = nil`, but `rawTypeString` is not `nullable`

Should it be `nullable`? Do we want to add an assert, that they never set `type = .unknown`?  Something else?

(It looks like this bug also exists in `STPSourceParams`, which I know I said would be a good model... sorry! Can you update that class too?)

Reply:
Setting `type = .unknown` seems like user error to me, an assertion against that sounds ok.  

Come to think of it, why is `type` publicly writable anyways?

Reply:
Great question! I suspect if it was `readonly`, that'd work.

It's possible to create a `SourcesParam` using just alloc/init, and then manually setting any fields needed. The static constructors make it really easy to do the right thing for the common case, for all the various types.

The combination of the `rawTypeString` and `additionalAPIParameters` property allow full control over the Source/PaymentMethod being created, since we don't necessarily keep up with every possible change to available sources/paymentmethods.

Plausibly there's a use case for building a `type` that the SDK knows about, but one where the static constructor interferes, in which case being able to set the `type` using an enum *might* be nice. I think that's a sufficiently rare case that `readonly` is 👍. It'd be a breaking change on SourceParams, but I think that's probably fine.

Reply:
Always a fan of more readonlys :)

Reply:
Looks like the case described actually happens in the SDK already for `STPSourceParams`:
```
            STPSourceParams *params = [STPSourceParams new];
            params.type = STPSourceTypeCard;
            params.token = token.tokenId;
```

I'm thinking I'll make `type` readonly for `STPPaymentMethodParams` but add an assertion for `STPSourceParams`.

Reply:
Looking at the tests, it really seems we intended to accommodate setting this value to Unknown.  e.g. one test does `XCTAssertNil(sourceParams.rawTypeString);`.  I'm inclined now to instead mark `rawTypeString` as `nullable` on `STPSourceParams`.